### PR TITLE
Adding percentage-based index management

### PR DESCRIPTION
**Pull request by btaani**
### Description
This PR adds the functionality of deleting old indices based on the used disk space of ES nodes.

/cc @sasagarw @shwetaap 

### Links
- JIRA: [LOG-1546](https://issues.redhat.com/browse/LOG-1546) and [LOG-2129](https://issues.redhat.com/browse/LOG-2129)
- Enhancement Proposal: https://github.com/openshift/enhancements/blob/master/enhancements/cluster-logging/elasticsearch-percentage-index-management.md
- Implementation details: https://docs.google.com/document/d/1Wtaa3ZWC1kQYAg_xu5YThlLtSZedgxQVvNAWZMaiSWw/edit#

### CODE REVIEW

File: internal/indexmanagement/reconcile.go [hunk 1]
```diff
@@ -362,6 +362,7 @@ func (imr *IndexManagementRequest) reconcileIndexManagementCronjob(policy apis.I
 		envvars = append(envvars,
 			corev1.EnvVar{Name: "MIN_AGE", Value: strconv.FormatUint(minAgeMillis, 10)},
 			corev1.EnvVar{Name: "NAMESPACE_SPECS", Value: namespaceSpecsString},
+			corev1.EnvVar{Name: "DISK_THRESHOLD", Value: strconv.FormatInt(policy.Phases.Delete.DiskThresholdPercent, 10)},
```
Comment (sasagarw) [inline internal/indexmanagement/reconcile.go:365]: Can we please do some validation here on `policy.Phases.Delete.DiskThresholdPercent`. For example, the % value should be between 0-100 otherwise return an error.

File: internal/indexmanagement/scripts.go [hunk 2]
```diff
@@ -131,6 +131,45 @@ def updateWriteIndex(currentIndex, nextIndex, alias):
...
+def getCurrentDiskPercent():
+  original_stdout = sys.stdout
+  try:
+    es_client = getEsClient()
+    response = es_client.cat.allocation(format="JSON", h="disk.percent,node", v=True)
+    diskPercent = []
+    for i in range(len(response)):
+      diskPercent.append(int(response[i]["disk.percent"]))
+    currentDiskPercent = max(diskPercent)
+    return currentDiskPercent
```
Comment (sasagarw) [inline internal/indexmanagement/scripts.go:144]: We can omit the `currentDiskPercent` variable here.
Directly `return max(diskPercent)`

File: internal/indexmanagement/scripts.go [hunk 3]
```diff
@@ -131,6 +131,45 @@ def updateWriteIndex(currentIndex, nextIndex, alias):
...
+    return False
+
+def deleteByPercentage(alias):
+  original_stdout = sys.stdout
+  try:
+    es_client = getEsClient()
+    indices = es_client.cat.indices(index=alias+"*",format="json",h="index,creation.date")
+
+    currentDiskPercent = getCurrentDiskPercent()
+    diskThreshold = os.environ['DISK_THRESHOLD']
```
Comment (sasagarw) [inline internal/indexmanagement/scripts.go:158]: Lets pass `diskThreshold` as a variable to `deleteByPercentage()`
What I mean is `def deleteByPercentage(alias, diskThreshold):`

File: internal/indexmanagement/scripts.go [hunk 4]
```diff
@@ -131,6 +131,45 @@ def updateWriteIndex(currentIndex, nextIndex, alias):
...
+    
+    while (int(currentDiskPercent) >= int(diskThreshold)):
```
Comment (sasagarw) [inline internal/indexmanagement/scripts.go:160]: `currentDiskPercent` is already integer. No need to typecast IMO.

Comment (sasagarw) [inline internal/indexmanagement/scripts.go:160]: Why a while loop over here? It might take time for ES to delete the indices and free up the disk usage.
IMO we should just run once and delete a single index. If the disk usage hasn't reached below threshold still then it will be done when the cron job runs next time. WDYT?

File: internal/indexmanagement/scripts.go [hunk 5]
```diff
@@ -131,6 +131,45 @@ def updateWriteIndex(currentIndex, nextIndex, alias):
...
+      print ("disk space threshold is exceeded, deleting the oldest index")
```
Comment (sasagarw) [inline internal/indexmanagement/scripts.go:161]: Let us also print in logs which index we are deleting

File: internal/indexmanagement/scripts.go [hunk 6]
```diff
@@ -131,6 +131,45 @@ def updateWriteIndex(currentIndex, nextIndex, alias):
...
+      if not indices:
+        break
+      oldest_index = min(indices, key=lambda s: s["creation.date"])
+      response = es_client.indices.delete(index=oldest_index['index'])
+      currentDiskPercent = getCurrentDiskPercent()
+  except Exception as e:
+    sys.stdout = open('/tmp/response.txt', 'w')
+    print(e)
+    sys.stdout = original_stdout
+
```
Comment (sasagarw) [inline internal/indexmanagement/scripts.go:172]: Do we want to return True or False over here?

**Review by sasagarw — CHANGES_REQUESTED** [review]
Good start :rocket: 
Requesting few changes which I could figure out at this point.

File: internal/indexmanagement/scripts.go (hunk 4 above)
Comment (btaani) [reply internal/indexmanagement/scripts.go:160]: The way I looked at it is to free up space as soon as possible before more data is indexed between the two rounds of cron jobs, which might be an issue if the jobs don't run as frequently as every 15 mins like in our examples

File: internal/indexmanagement/scripts.go (hunk 6 above)
Comment (btaani) [reply internal/indexmanagement/scripts.go:172]: Yes you're right.

File: internal/indexmanagement/scripts.go (hunk 4 above)
Comment (sasagarw) [reply internal/indexmanagement/scripts.go:160]: But the thing is does ES free up the disk space as soon as you delete the index? Can you test this case and verify once?
I am telling so because I feel ES takes some time to free up the space. If that happens then because of running while loop we might delete too many indices.

Comment (btaani) [reply internal/indexmanagement/scripts.go:160]: You are right, I totally forgot about this. I'll update the code

File: internal/indexmanagement/scripts.go [hunk 7]
```diff
@@ -131,6 +131,45 @@ def updateWriteIndex(currentIndex, nextIndex, alias):
...
+def getCurrentDiskPercent():
+  original_stdout = sys.stdout
+  try:
+    es_client = getEsClient()
+    response = es_client.cat.allocation(format="JSON", h="disk.percent,node", v=True)
+    diskPercent = []
+    for i in range(len(response)):
+      diskPercent.append(int(response[i]["disk.percent"]))
+    return max(diskPercent)
```
Comment (periklis) [inline internal/indexmanagement/scripts.go:143]: I believe the approach by calculating the max over the collected disk space usage per node can be simplified by asking ES through the nodes stats API for the FS information: https://www.elastic.co/guide/en/elasticsearch/reference/6.8/cluster-nodes-stats.html#fs-info

Comment (btaani) [reply internal/indexmanagement/scripts.go:143]: I do not see how it can be simplified, the stats API that you linked gives the following info for each node, here is a snapshot from my cluster:

> "fs" : {
        "timestamp" : 1646828572706,
        "total" : {
          "total_in_bytes" : 128300593152,
          "free_in_bytes" : 115112910848,
          "available_in_bytes" : 115112910848
        },
when I calculate the disk percentage using those values, I get the same result that the cat allocation API gives me without any calculations.
I end up with 3 values in both APIs, one for each node.

Comment (periklis) [reply internal/indexmanagement/scripts.go:143]: @btaani You can ignore my comment. I accidentally thought that the `nodes/stats/fs` call will result in a total consumption output across all nodes.

**Comment by periklis** [comment]
Based on our discussion last week, I am providing a short summary of the topics we want to address moving forward and to keep a record:

## Identified issues with the current approach
The current approach taken leverages the space consumption of the ES node with the highest consumption so far. In addition it deletes an index at each run without a clear indication if the `DiskPercentage` threshold is satisfied again. In addition it does not consider the OpenShift Logging indices modesl (i.e. `app-*`, `infra-*`, `audit-*`) to identify oldest indices with the highest impact on disk consumption across nodes.

## Topics we need to consider to improve the implementation
1. The cronjob should consider to create first a valid snapshot of what is the max allowed space across all nodes (e.g.[lightweight-curator](https://github.com/Tessg22/lightweight-curator/blob/master/scripts/lightweight_curator.py#L99:5))
2. The cronjob should leverage the OpenShift Logging model and create a list of oldest indices which when deleted can satisfy the `DiskPercentage` threshold (e.g. [lightweight-curator](https://github.com/Tessg22/lightweight-curator/blob/master/scripts/lightweight_curator.py#L133))
3. How do we prevent over-deletion or too eager deletion in general? This is a question more to how the ES API handles deleted indices. In general ES releases FS space in the background. This raises the question if deleted indices but not yet released from the filesystem play any role in the `max allowed space` calculation.
4. We need to consider that the `PruneNamespaces` job runs in parallel and in turn tries to delete single documents across namespaces. Is the present implementation of this job fault-tolerant enough to consider the case trying to delete a document on an index that was immediately deleted by the `DiskPercentage` threshold routines?

File: internal/indexmanagement/scripts.go [hunk 8]
```diff
@@ -131,6 +131,92 @@ def updateWriteIndex(currentIndex, nextIndex, alias):
...
+  # Returns the total disk space of all nodes combined multiplied by the pre-defined threshold
+  original_stdout = sys.stdout
+  try:
+    es_client = getEsClient()
+    response = es_client.cat.allocation(format="JSON", h="disk.total", bytes="kb")
+    totalSize = 0
+    for i in range(3):
+      #print (response[i]["disk.total"])
+      totalSize = totalSize + int(response[i]["disk.total"])
+    return int(diskThreshold/100.0 * totalSize)
```
Comment (periklis) [inline internal/indexmanagement/scripts.go:145]: Why `range(3)`? A cluster can have more than three nodes.

File: internal/indexmanagement/scripts.go [hunk 9]
```diff
@@ -131,6 +131,92 @@ def updateWriteIndex(currentIndex, nextIndex, alias):
...
+      print (index['index'], "'s size is starting to get over threshold")
+      if index['index'].startswith(alias):
+        print ("----------", index['index'], "should be deleted")
+        indicesToDelete.append(index['index'])
+    return indicesToDelete, indicesSizeCounter
+  except Exception as e:
+    #sys.stdout = open('/tmp/response.txt', 'w')
+    print(e)
+    sys.stdout = original_stdout
+    return -1
```
Comment (periklis) [inline internal/indexmanagement/scripts.go:173]: What if we flip the traversal here from oldest to newest? This should simplify things a lot, because by default we append indices (starting from oldest) until the sum of the appended indices hits the threshold. WDYT?

Comment (btaani) [reply internal/indexmanagement/scripts.go:173]: After doing a few more tests, I figured that the current way is the only one to prevent over-deletion of indices.
Earlier, I saw some indices being deleted each time the cron job was running. That actually wasn't over-deletion, since we are counting up the index sizes from newest to oldest, until the size reaches the threshold. So each time an old index was deleted, it was because the threshold was actually exceeded since new indices were also being created (due to rollover), thus taking up more space.
For example, if we have 100GB of total disk space for all nodes, and the threshold is 70%, we add up index sizes starting from the most recent index. When the sum exceeds 70GB, we start deleting indices, which would be the oldest ones of course. 
In conclusion, the way that lightweight-curator was implemented in the first place was to prevent over-deletion.

**Review by sasagarw — CHANGES_REQUESTED** [review]
Just one nit

File: internal/indexmanagement/reconcile.go [hunk 10]
```diff
@@ -359,9 +359,16 @@ func (imr *IndexManagementRequest) reconcileIndexManagementCronjob(policy apis.I
...
+		diskThreshold := policy.Phases.Delete.DiskThresholdPercent
+		if diskThreshold < 0 || diskThreshold > 100 {
+			err := fmt.Errorf("invalid percentage value")
+			return kverrors.Wrap(err, ", %d should be between 0 and 100", diskThreshold)
+		}
+
 		envvars = append(envvars,
 			corev1.EnvVar{Name: "MIN_AGE", Value: strconv.FormatUint(minAgeMillis, 10)},
 			corev1.EnvVar{Name: "NAMESPACE_SPECS", Value: namespaceSpecsString},
+			corev1.EnvVar{Name: "DISK_THRESHOLD", Value: strconv.FormatInt(policy.Phases.Delete.DiskThresholdPercent, 10)},
```
Comment (sasagarw) [inline internal/indexmanagement/reconcile.go:371]: We can use the `diskThreshold` variable itself here

File: internal/indexmanagement/scripts.go [hunk 11]
```diff
@@ -131,6 +131,83 @@ def updateWriteIndex(currentIndex, nextIndex, alias):
...
+
+def deleteByPercentage(alias, diskThreshold):
+  original_stdout = sys.stdout
+  try:
+    es_client = getEsClient()
+    
+    # Get the maximum allowed size, that if exceeded, old indices are deleted
+    maxAllowedSize = getMaxAllowedSize(int(diskThreshold))
```
[the remark this answers is not in the document]
Comment (btaani) [reply internal/indexmanagement/scripts.go:195]: This `diskThreshold` is the one coming from the shell script in [line 548](https://github.com/openshift/elasticsearch-operator/blob/b5ca93ab6be6e6230b08f11adb6d53aa393dd39e/internal/indexmanagement/scripts.go#L548) below, so at this point it's still a string, this is why I am casting it to `int` to use it in the `maxAllowedSize` calculation

File: internal/indexmanagement/reconcile.go (hunk 10 above)
Comment (sasagarw) [reply internal/indexmanagement/reconcile.go:371]:
```suggestion
			corev1.EnvVar{Name: "DISK_THRESHOLD", Value: strconv.FormatInt(diskThreshold, 10)},
```

File: internal/indexmanagement/scripts.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (sasagarw) [reply internal/indexmanagement/scripts.go:195]: Okay. My mistake in reading.

File: internal/indexmanagement/scripts.go [hunk 12]
```diff
@@ -131,6 +131,83 @@ def updateWriteIndex(currentIndex, nextIndex, alias):
...
+  # maxAllowedSize, start adding indices to the list
+  original_stdout = sys.stdout
+  try:
+    expectedSize = int(index['store.size']) + indicesSizeCounter
+    if indicesSizeCounter < maxAllowedSize and expectedSize < maxAllowedSize:
+      indicesSizeCounter += int(index['store.size'])
+    else:
+      if index['index'].startswith(alias):
+        indicesToDelete.append(index['index'])
+    return indicesToDelete, indicesSizeCounter
```
Comment (Red-GV) [inline internal/indexmanagement/scripts.go:163]: Maybe assign `indicesSizeCounter` and `indicesToDelete` to different variables before modifying it? Seems kinda odd to modify the arguments like this.

File: internal/indexmanagement/scripts.go [hunk 13]
```diff
@@ -131,6 +131,83 @@ def updateWriteIndex(currentIndex, nextIndex, alias):
...
+
+def getDeletableIndices(index, alias, maxAllowedSize, indicesSizeCounter, indicesToDelete):
+  # Returns a list of indices from [alias] (infra, app, or audit), that should be deleted
+  # traverse through indices from newest to oldest, adding up their size. When the sum of sizes exceeds
+  # maxAllowedSize, start adding indices to the list
+  original_stdout = sys.stdout
+  try:
+    expectedSize = int(index['store.size']) + indicesSizeCounter
+    if indicesSizeCounter < maxAllowedSize and expectedSize < maxAllowedSize:
+      indicesSizeCounter += int(index['store.size'])
```
Comment (Red-GV) [inline internal/indexmanagement/scripts.go:159]: Maybe this instead? Is there ever a time where `expectedSize` < `indicesSizeCounter`?

```suggestion
    if expectedSize < maxAllowedSize:
      indicesSizeCounter += int(index['store.size'])
```

Comment (btaani) [reply internal/indexmanagement/scripts.go]: Thanks @Red-GV
I have updated the code with you suggestions + merged two functions into one (the ones that return the list of indices to delete)

**Review by Red-GV — COMMENTED** [review]
Few more nits and I think we're good.

File: internal/indexmanagement/scripts.go [hunk 14]
```diff
@@ -131,6 +131,72 @@ def updateWriteIndex(currentIndex, nextIndex, alias):
...
+def getDeletableIndices(index, alias, maxAllowedSize, indicesSizeCounter, indicesToDelete):
+  # Returns a list of indices from [alias] (infra, app, or audit), that should be deleted
+  # traverse through indices from newest to oldest, adding up their size. When the sum of sizes exceeds
+  # maxAllowedSize, start adding indices to the list
+  original_stdout = sys.stdout
+  try:
+    indicesSizeCounter = indicesSizeCounter
+    indicesToDelete = indicesToDelete
+    expectedSize = int(index['store.size']) + indicesSizeCounter
+    if indicesSizeCounter < maxAllowedSize:
```
[the remark this answers is not in the document]
Comment (btaani) [reply internal/indexmanagement/scripts.go:160]: Yes, I miss this one!
